visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp

```diff
@@ -0,0 +1,671 @@
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+/**
+ * @file cubical-complex-collapse.cpp
+ * @ingroup Examples
+ * @author Jacques-Olivier Lachaud (\c jacques-olivier.lachaud@univ-savoie.fr )
+ * Laboratory of Mathematics (CNRS, UMR 5127), University of Savoie, France
+ *
+ * @date 2015/08/28
+ *
+ * An example file named cubical-complex-collapse.cpp.
+ *
+ * This file is part of the DGtal library.
+ */
+
+///////////////////////////////////////////////////////////////////////////////
+#include <iostream>
+#include <map>
+#include <boost/program_options/options_description.hpp>
+#include <boost/program_options/parsers.hpp>
+#include <boost/program_options/variables_map.hpp>
+
+#include "DGtal/base/Common.h"
+#include "DGtal/base/CountedPtr.h"
+#include "DGtal/helpers/StdDefs.h"
+#include "DGtal/math/MPolynomial.h"
+#include "DGtal/io/readers/MPolynomialReader.h"
+#include "DGtal/io/DrawWithDisplay3DModifier.h"
+#include "DGtal/io/viewers/Viewer3D.h"
+#include "DGtal/topology/KhalimskySpaceND.h"
+#include "DGtal/topology/CubicalComplex.h"
+#include "DGtal/topology/CubicalComplexFunctions.h"
+#include "DGtal/topology/SetOfSurfels.h"
+#include "DGtal/topology/DigitalSurface.h"
+#include "DGtal/topology/helpers/Surfaces.h"
+#include "DGtal/shapes/GaussDigitizer.h"
+#include "DGtal/shapes/Mesh.h"
+#include "DGtal/shapes/implicit/ImplicitPolynomial3Shape.h"
+
+///////////////////////////////////////////////////////////////////////////////
+
+using namespace std;
+using namespace DGtal;
+
+
+/**
+ * This 4D extension of a 3D implicit function transforms f(x,y,z) as:
+ * \f$ F(x,y,z,t) = f(x,y,z) - t | \nabla f(x,y,z) | \f$.
+ */
+template <typename TSpace3, typename TSpace4, typename TPolynomial3>
+struct ImplicitSurface4DProductExtension {
+  typedef TSpace3                          Space3;
+  typedef TSpace4                          Space4;
+  typedef TPolynomial3                     Polynomial3;
+  typedef typename Space3::RealPoint       RealPoint3;
+  typedef typename Space3::RealVector      RealVector3;
+  typedef typename Space3::Integer         Integer;
+  typedef typename RealPoint3::Coordinate  Ring;
+  typedef typename Space4::RealPoint       RealPoint4;
+  typedef typename Space4::RealVector      RealVector4;
+  typedef RealPoint4                       RealPoint;
+  typedef RealVector4                      RealVector;
+
+  ImplicitSurface4DProductExtension( Clone<Polynomial3> poly )
+    : f( poly )
+  {
+    fx  = derivative<0>( f );
```

Comment:
it comes from the DEC class but I d'ont see the include .. (I see perhaps DrawWithDisplay3DModifier.h)